src/System.Threading/tests/MonitorTests.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Xunit;
+
+namespace System.Threading.Tests
+{
+    static class MonitorTests
+    {
+        // Attempts a single recursive acqusition/release cycle of a newly-created lock.
+        [Fact]
+        public static void BasicRecursion(ref string message)
+        {
+            var obj = new object();
+            Assert.True(Monitor.TryEnter(obj));
+            Assert.True(Monitor.TryEnter(obj));
+            Monitor.Exit(obj);
+            Assert.True(Monitor.IsEntered(obj));
+            Monitor.Enter(obj);
+            Assert.True(Monitor.IsEntered(obj));
+        }
+
+        // Attempts to overflow the recursion count of a newly-created lock.
+        [Fact]
+        public static void DeepRecursion(ref string message)
+        {
+            var obj = new object();
+            var hc = obj.GetHashCode();
+            // reduced from "(long)int.MaxValue + 2;" to something that will return in a more meaningful time
+            const int limit = 10000;
+
+            for (var i = 0L; i < limit; i++)
+            {
+                Assert.True(Monitor.TryEnter(obj));
+            }
+
+            for (var j = 0L; j < (limit - 1); j++)
+            {
+                Monitor.Exit(obj);
+                Assert.True(Monitor.IsEntered(obj));
+            }
+
+            Monitor.Exit(obj);
+            Assert.True(Monitor.IsEntered(obj));
+        }
+
+        [Fact]
+        public static void IsEntered()
+        {
+            var obj = new object();
+            Assert.False(Monitor.IsEntered(obj));
+            lock (obj)
+            {
+                Assert.True(Monitor.IsEntered(obj));
+            }
+            Assert.False(Monitor.IsEntered(obj));
+        }
+
+        [Fact]
+        public static void Enter_Invalid()
+        {
+            bool lockTaken = false;
+            var obj = new object();
+
+            Assert.Throws<ArgumentNullException>(() => Monitor.Enter(null));
+            Assert.Throws<ArgumentNullException>(() => Monitor.Enter(null, ref lockTaken));
```

Comment:
For all of the Throws for argument exceptions, can you add the parameter name we expect to be in the exception?